### DISCUSSION TITLE
WARNING Some velocities at the rotor are negative.

### DISCUSSION THREAD

**Discussion by nrz22** [NatLabRockies/floris#961] [Q&A] [answered]
Hello, I have been trying to model and calculate the AEP for two wind farms near each other with a few different layouts, and when I run the model I encounter this warning, "WARNING Some velocities at the rotor are negative". The power output for each wind speed and wind direction is definitely not right but I'm not sure what my problem is. I have successfully been able to calculate for one wind farm on its own using a similar code. My original thought was that because I was modeling two wind farms aligned east to west, but by combining the coordinates into one array with the western-most first and the eastern-most, that maybe the order of the turbines was messing with the model, but then I tried again with two farms aligned north to south and still got the error. Any help on this issue would be much appreciated.

**Comment by nrz22** [comment]
This is my code 

location_params = {
    'NW': {
        'Power': np.array([10]),#, 10, 10, 15, 15, 15,
        'SF': np.array([8]),#, 9, 10, 8, 9, 10]),
        'Dir': ['east'],#, 'east', 'east', 'east', 'east', 'east'],
        'Skew': np.array([1.4]),#, 2.4, 0.4, 1.2, 2.4, 1]),
        'Shrink': np.array([1]),#, 0.6, 0.5, 1, 0.6, 0.5]),
        'Shift': np.array([0]),#, 3.5, 4.5, 0, 3.7, 5])
    },
    'NE': {
        'Power': np.array([10]),#, 10, 10, 15, 15, 15]),
        'SF': np.array([8]),#, 9, 10, 8, 9, 10]),
        'Dir': ['west'],#, 'west', 'west', 'west', 'west', 'west'],
        'Skew': np.array([1.4]),#, 1.2, 0.4, 1.4, 2.4, 0.2]),
        'Shrink': np.array([1]),#, 0.8, 0.5, 1, 0.6, 0.5]),
        'Shift': np.array([0]),#, 1.5, 5, 0, 3.55, 5])


def run_simulation(combination, farm_coords_list, turbine_sizes, location_params, model_name):
    # Generate combined wind farm layout
    
    cluster = Wind_Farm.wf_super(farm_coords_list[0], farm_coords_list[1])
    
    # Sort farm_coords_list and turbine_sizes based on Y-coordinates (north to south)
    sorted_farm = sorted(zip(cluster, turbine_sizes), key=lambda p: p[0][1], reverse=True)
    farm_coords_list, turbine_sizes = zip(*sorted_farm)

    combined_x_coords, combined_y_coords = [point[0] for point in farm_coords_list], [point[1] for point in farm_coords_list]
    # print(combined_x_coords, combined_y_coords, turbine_sizes)
    
    # Use the first location name from location_params to retrieve wind stats
    location_name = list(location_params.keys())[0]
    location_obj = predefined_wind_farms[location_name]
    wind_stats = Wind_Stats(location_obj)
    
    # Extract the relevant parameters for the first location in the combination
    loc_params = location_params[location_name]
    power_value = loc_params['Power'][combination[0]]
    
    # Generate wind distribution and power curves
    A, k, freq, wd_array = wind_stats.wind_dist(height=wind_stats.power_curves(power_value)[2])
    wind_speeds, powers, height = wind_stats.power_curves(power_value)
    ws_array, powers, TI_array = wind_stats.interp_wind(wind_speeds, powers, location_obj.Uref, height)
    weibull_values = wind_stats.weib_prob(A, k, ws_array)
    hours = wind_stats.hours_prob(weibull_values)
    
    
    # Run FLORIS model
    fmodel = get_model(model_name, power_value)
    print(model_name, power_value)
    fmodel.set(layout_x=combined_x_coords, layout_y=combined_y_coords)
    
    F1_Power = []
    F2_Power = []
    for wd in wd_array:
        wind_data_wd_sweep = TimeSeries(
            wind_speeds=ws_array,
            wind_directions=float(wd),
            turbulence_intensities=TI_array
        )
        fmodel.set(wind_data=wind_data_wd_sweep)
        fmodel.run()
        
        # Get power data for all turbines
        power_data = fmodel.get_turbine_powers()
        print(power_data)
        


       # Main execution function
def main():
    start_time = time.time()
    directory = "C:/Users/"
    os.makedirs(directory, exist_ok=True)
    
    excel_file_name = os.path.join(directory, f".xlsx")
    # Check if the file exists
    if not os.path.exists(excel_file_name):
        # Create a new Excel file
        wb = Workbook()
        wb.save(excel_file_name)

    model_data = {model_name: [] for model_name in model_names}

    # Locations for north and south clusters
    north_sites = ['NW', 'NE']
    #south_sites = ['SW', 'SE']

    # Iterate over north and south clusters
    for cluster_name, cluster_locations in [('North', north_sites)]:#, ('South', south_sites)]:
        for combination in product(*[range(len(location_params[loc]['Power'])) for loc in cluster_locations]):
            x_list = []
            y_list = []
            farm_coords_list = []
            turbine_sizes = []

            # Generate the layout for each location in the cluster
            for idx, location_name in enumerate(cluster_locations):
                loc_params = location_params[location_name]
                location_obj = predefined_wind_farms[location_name]

                # Setting the parameters from the current combination
                location_obj.site_layout(
                    power=loc_params['Power'][combination[idx]],
                    spacing_factor=loc_params['SF'][combination[idx]],
                    corner=location_name.split('_')[1],
                    direction=loc_params['Dir'][combination[idx]],
                    offset=0,  # Assuming no offset for simplicity
                    skew=loc_params['Skew'][combination[idx]],
                    shrink_x=loc_params['Shrink'][combination[idx]],
                    shift=loc_params['Shift'][combination[idx]],
                    show_legend=False
                )
                plt.show
            
                
                farm_coords_list.append(location_obj.turbine_locations)
                # turbine_sizes.append([loc_params['Power'][combination[idx]]] * len(location_obj.turbine_locations))
                turbine_sizes.extend([loc_params['Power'][combination[idx]]] * len(location_obj.turbine_locations))

**Comment by misi9170** [comment, marked as answer]
Hi @nrz22 , this usually happens when turbines are placed close together, so that a turbine may be in the near-wake zone of another turbine. Floris does not model the near wake very well, because turbines are rarely placed in such close proximity. Would you be able to share `fmodel.layout_x` and fmodel.layout_y` for your case? If there is close spacing (say, less than 3 rotor diameters) between any of the turbines, this could be what is causing the negative velocities.

**Reply by nrz22** [reply]
Hello, thanks for your quick response. 

I had seen in FlorisModel that that was one possible reason so I had already checked, but it was with a plot that worked in km, but I had missed the conversion to m in my coordinates for the model set. So, the issue is resolved, thanks.

**Reply by misi9170** [reply]
Great, glad to hear it!
